# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

In [1]:
import pandas as pd

## imports for Python, Pandas

In [2]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [126]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]
data;

In [124]:
# use normalization to create tables from nested element
json_normalize(data, 'counties');

In [123]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']]);

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [122]:
# load json as string
json.load((open('data/world_bank_projects_less.json')));

In [127]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')
sample_json_df;

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

## Loading data and getting information about the data

In [8]:
# the json file read by using the pandas
df = pd.read_json('data/world_bank_projects.json') 
# get the general information about the data file
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 50 columns):
_id                         500 non-null object
approvalfy                  500 non-null int64
board_approval_month        500 non-null object
boardapprovaldate           500 non-null object
borrower                    485 non-null object
closingdate                 370 non-null object
country_namecode            500 non-null object
countrycode                 500 non-null object
countryname                 500 non-null object
countryshortname            500 non-null object
docty                       446 non-null object
envassesmentcategorycode    430 non-null object
grantamt                    500 non-null int64
ibrdcommamt                 500 non-null int64
id                          500 non-null object
idacommamt                  500 non-null int64
impagency                   472 non-null object
lendinginstr                495 non-null object
lendinginstrtype            495 non

Data contains total 500 entries with some columns with null values. It means entries corresponding to few columns are missing. But, the countryname and project_names are with no missinig values. 

In [9]:
#Count the entries using the column 'countryname'.
df_values = df.countryname.value_counts()
#Display the most 10 values
df_sorted = df_values.head(10)
df_sorted

Republic of Indonesia              19
People's Republic of China         19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
People's Republic of Bangladesh    12
Nepal                              12
Kingdom of Morocco                 12
Africa                             11
Republic of Mozambique             11
Name: countryname, dtype: int64

## Loading the flat data using json.load() and getting the project themes: 

In [10]:
#Read the data file using json.load((open('')))
df = json.load((open('data/world_bank_projects.json')))
#Tabulate the required data, dictionary key and values (code and name) of the mjtheme_codename
df_normal= json_normalize(df, 'mjtheme_namecode')
df_normal.head()

code                                   name
0    8                      Human development
1   11                                       
2    1                    Economic management
3    6  Social protection and risk management
4    5                  Trade and integration

Here, we see there are name of the project themes are missing for a lot of entries.

In [114]:
#sort the values using .sort_vaues()
df_normal_sorted = df_normal.sort_values(['code', 'name'])
df_normal_sorted.head(10)

code                 name
212     1                     
363     1                     
1024    1                     
1114    1                     
1437    1                     
2       1  Economic management
88      1  Economic management
175     1  Economic management
204     1  Economic management
205     1  Economic management

Missed names corresponding to the particular codes are filled using back filling 'bfill' method. '.fillna()' requires the blanc spaces to be replaced by 'nan'

In [118]:
#replace the blanc spaces by 'nan'
import numpy as np
df_normal_sorted.name[df_normal_sorted['name']== '']  = np.nan
df_normal_sorted.head(10)

code                 name
212     1                  NaN
363     1                  NaN
1024    1                  NaN
1114    1                  NaN
1437    1                  NaN
2       1  Economic management
88      1  Economic management
175     1  Economic management
204     1  Economic management
205     1  Economic management

###  DataFrame with the missing data filled 

In [121]:
#fill the nan cells using back filling method. 
df_filled = df_normal_sorted.fillna(method='bfill')
df_filled.head(10)

code                 name
212     1  Economic management
363     1  Economic management
1024    1  Economic management
1114    1  Economic management
1437    1  Economic management
2       1  Economic management
88      1  Economic management
175     1  Economic management
204     1  Economic management
205     1  Economic management

## To find the top 10 project themes:

Now, we sort the values using the name column, which sorts the values and tabulates in the ascending order

In [133]:
df2 = df_filled.name.value_counts().head(10)
df2

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Name: name, dtype: int64

In [128]:
### End of codes###

# Summary




### 1) Names of countries with the 10 most projects:

|Name of Country |$\;\;\;\;\;\;\;\;\;\;\;\;\;$        Number of Projects|
|:-:   |  :-: |
|People's Republic of China                     |19|
|Republic of Indonesia                           |   19|
|Socialist Republic of Vietnam                  |   17|
|Republic of India                               |   16|
|Republic of Yemen                               |   13|
|Nepal                                           |   12|
|People's Republic of Bangladesh       |   12|
|Kingdom of Morocco                   |   12|
|Republic of Mozambique               |   11|
|Africa                            |   11|



### 2) The top 10 project themes: 

|Project Themes$\;\;\;\;\;\;\;\;\;\;\;\;\;$|Counts|
| :---: |     :---:        |
|Environment and natural resources management    |250|
|Rural development                               |216|
|Human development                               |210|
|Public sector governance                        |199|
|Social protection and risk management           |168|
|Financial and private sector development        |146|
|Social dev/gender/inclusion                     |130|
|Trade and integration                           | 77|
|Urban development                               | 50|
|Economic management                             | 38|

##### 3) During the prcess of finding the top 10 projects we found the missing entries for the names of the projects. By sorting the data with respect to the code and the name of the project, all the empty cells are appeared at the beginning of the name columns for each 'code'. We filled the empty spaces with the back filling method. we have shown the 10 rows of the filled dataframe in the above codelines.  